File: tooling/runtimecheck/state.go
package runtimecheck

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	goruntime "runtime"
	"sort"

	"github.com/d7z-team/mini-go/compiler/target"
	"github.com/d7z-team/mini-go/runtime"
	"github.com/d7z-team/mini-go/runtime/bytecode"
)

// StateVector records exact owner observations for minimal bytecode inputs.
type StateVector struct {
	Name                 string                  `json:"name"`
	Image                bytecode.ExecutionImage `json:"image"`
	Limit                int64                   `json:"limit"`
	Actions              []StateAction           `json:"actions"`
	InitializationError  string                  `json:"initialization_error,omitempty"`
	InitializationSteps  int64                   `json:"initialization_steps"`
	InitializationMemory [4]int64                `json:"initialization_memory"`
}

type StateAction struct {
	Operation string                 `json:"operation"`
	State     runtime.ExecutionState `json:"state,omitempty"`
	Error     string                 `json:"error,omitempty"`
	Steps     int                    `json:"steps"`
	Memory    [4]int64               `json:"memory"`
}

// GenerateStateVectors runs repeated invocations without compiling source or
// clearing instance state. Every observation comes from the current Go owner.
func GenerateStateVectors(ctx context.Context, inputs map[string][]byte) ([]byte, error) {
	names := make([]string, 0, len(inputs))
	for name := range inputs {
		names = append(names, name)
	}
	sort.Strings(names)
	var vectors []StateVector
	for _, name := range names {
		var artifacts []bytecode.Artifact
		input := bytes.TrimSpace(inputs[name])
		if len(input) > 0 && input[0] == '[' {
			if err := json.Unmarshal(input, &artifacts); err != nil {
				return nil, fmt.Errorf("%s: %w", name, err)
			}
		} else {
			artifacts = make([]bytecode.Artifact, 1)
			if err := json.Unmarshal(input, &artifacts[0]); err != nil {
				return nil, fmt.Errorf("%s: %w", name, err)
			}
		}
		if len(artifacts) == 0 {
			return nil, fmt.Errorf("%s: missing root artifact", name)
		}
		module := artifacts[0].Module.Path
		image := bytecode.ExecutionImage{
			Format: bytecode.ExecutionFormat, Version: bytecode.ExecutionVersion,
			CompilerID: bytecode.CompilerIdentity, ContractID: bytecode.ExecutionContract,
			Root: module, Target: target.Target{Tags: []string{"minigo"}},
			Entries:  []bytecode.Entry{{Name: "default", ModulePath: module, FunctionID: "fn.Main"}},
			Packages: make(map[string]bytecode.PackageArchive),
		}
		for _, artifact := range artifacts {
			artifact.Format, artifact.Version, artifact.OpcodeSet = bytecode.Format, bytecode.CurrentVersion, bytecode.OpcodeSet
			hash, err := bytecode.Hash(&artifact)
			if err != nil {
				return nil, err
			}
			encoded, err := json.Marshal(artifact)
			if err != nil {
				return nil, err
			}
			if _, exists := image.Packages[artifact.Module.Path]; exists {
				return nil, fmt.Errorf("%s: duplicate module %s", name, artifact.Module.Path)
			}
			image.Packages[artifact.Module.Path] = bytecode.PackageArchive{Artifact: encoded, ArtifactHash: hash}
		}
		var err error
		image.Hash, err = bytecode.HashExecutionImage(image)
		if err != nil {
			return nil, err
		}
		program, err := runtime.LoadExecutionImage(image)
		if err != nil {
			return nil, err
		}
		for _, limit := range []int64{128, 159, 160, 512, 600, 640, 1024, 1536, 2048, 4096} {
			if err := ctx.Err(); err != nil {
				return nil, err
			}
			vector := StateVector{Name: name, Image: image, Limit: limit, Actions: []StateAction{}}
			if err := func() error {
				instance, err := program.Instantiate(ctx, runtime.InstanceOptions{Limits: runtime.Limits{MaxAllocatedBytes: limit, MaxSteps: 1000}})
				if err != nil {
					if err := ctx.Err(); err != nil {
						return err
					}
					projected, projectionErr := runtime.ProjectExecutionError(err)
					if projectionErr != nil {
						return projectionErr
					}
					vector.InitializationError = projected.Code
					return nil
				}
				defer instance.Close()
				initial, err := instance.RuntimeStats(ctx)
				if err != nil {
					return err
				}
				vector.InitializationSteps = initial.ExecutedSteps
				vector.InitializationMemory = [4]int64{initial.LiveGuestBytes, initial.AllocatedSinceSweep, initial.TotalAllocatedBytes, initial.PeakGuestBytes}
				observe := func(action StateAction, operationErr error) error {
					if operationErr != nil {
						projected, err := runtime.ProjectExecutionError(operationErr)
						if err != nil {
							return err
						}
						action.Error = projected.Code
					}
					stats, err := instance.RuntimeStats(ctx)
					if err != nil {
						return err
					}
					action.Memory = [4]int64{stats.LiveGuestBytes, stats.AllocatedSinceSweep, stats.TotalAllocatedBytes, stats.PeakGuestBytes}
					vector.Actions = append(vector.Actions, action)
					return nil
				}
				for invocation := range 4 {
					execution, err := instance.Start("default")
					if observeErr := observe(StateAction{Operation: "start"}, err); observeErr != nil {
						return observeErr
					}
					if err != nil {
						continue
					}
					for step := 0; step < 1001; step++ {
						state, count, err := execution.PollSteps(1)
						// The cancellation supervisor may briefly own the VM. Busy
						// performs no guest work and is not a bytecode observation.
						for errors.Is(err, runtime.ErrBusy) {
							if err := ctx.Err(); err != nil {
								return err
							}
							goruntime.Gosched()
							state, count, err = execution.PollSteps(1)
						}
						if observeErr := observe(StateAction{Operation: "poll", State: state, Steps: count}, err); observeErr != nil {
							return observeErr
						}
						if err != nil || state == runtime.ExecutionCompleted {
							break
						}
						if invocation == 2 {
							execution.Cancel()
							if err := observe(StateAction{Operation: "cancel"}, nil); err != nil {
								return err
							}
							break
						}
						if step == 1000 {
							return fmt.Errorf("%s did not terminate within its instruction limit", name)
						}
					}
				}
				return nil
			}(); err != nil {
				return nil, fmt.Errorf("%s limit %d: %w", name, limit, err)
			}
			vectors = append(vectors, vector)
		}
	}
	return json.Marshal(vectors)
}

File: playground/runtime-rust/tests/state.rs
use mini_go::{
    instance::{ExecutionLimits, Instance, PollStatus},
    loader::LoadLimits,
    program::Program,
};
use serde::Deserialize;
use serde_json::value::RawValue;
use std::sync::Arc;

#[derive(Deserialize)]
struct Vector {
    name: String,
    image: Box<RawValue>,
    limit: u64,
    actions: Vec<Action>,
    #[serde(default)]
    initialization_error: String,
    initialization_steps: u64,
    initialization_memory: [u64; 4],
}

#[derive(Deserialize)]
struct Action {
    operation: String,
    #[serde(default)]
    state: String,
    #[serde(default)]
    error: String,
    steps: u64,
    memory: [u64; 4],
}

#[test]
fn owner_actions_and_memory_match_current_go_observations() {
    let vectors: Vec<Vector> =
        serde_json::from_str(include_str!("../../../testdata/runtime/state.json")).unwrap();
    assert!(!vectors.is_empty());
    for vector in vectors {
        let program =
            Arc::new(Program::load(vector.image.get().as_bytes(), LoadLimits::default()).unwrap());
        let mut instance = Instance::new(
            program,
            ExecutionLimits {
                max_allocated_bytes: vector.limit,
                max_steps: 1000,
                ..ExecutionLimits::default()
            },
        )
        .unwrap();
        let initialization = instance.initialize_root(&Default::default());
        if !vector.initialization_error.is_empty() {
            assert_eq!(
                initialization.unwrap_err().code,
                vector
                    .initialization_error
                    .strip_prefix("execution.")
                    .unwrap_or(&vector.initialization_error),
                "{} limit {} initialization",
                vector.name,
                vector.limit
            );
            assert_eq!(instance.heap_stats().live_objects, 0);
            continue;
        }
        initialization.unwrap();
        assert_eq!(
            instance.steps(),
            vector.initialization_steps,
            "{} initialization steps",
            vector.name
        );
        let initial = instance.memory_stats();
        assert_eq!(
            [
                initial.live_bytes,
                initial.allocated_since_sweep,
                initial.total_allocated_bytes,
                initial.peak_bytes
            ],
            vector.initialization_memory,
            "{} limit {} initialization memory",
            vector.name,
            vector.limit
        );
        for (index, action) in vector.actions.iter().enumerate() {
            let context = format!(
                "{} limit {} action {index} {}",
                vector.name, vector.limit, action.operation
            );
            let before = instance.steps();
            let result = match action.operation.as_str() {
                "start" => instance.start("default", vec![]).map(|()| String::new()),
                "cancel" => instance.cancel().map(|()| String::new()),
                "poll" => instance.poll_steps(1).map(|state| {
                    match state {
                        PollStatus::Running => "running",
                        PollStatus::Ready => "completed",
                        PollStatus::Pending => "blocked",
                        PollStatus::Paused => "paused",
                    }
                    .to_owned()
                }),
                operation => panic!("unknown owner action {operation}"),
            };
            match result {
                Ok(state) => {
                    assert!(
                        action.error.is_empty(),
                        "{context}: expected {}",
                        action.error
                    );
                    assert_eq!(state, action.state, "{context}");
                }
                Err(error) => assert_eq!(
                    error.code,
                    action
                        .error
                        .strip_prefix("execution.")
                        .unwrap_or(&action.error),
                    "{context}: {error}"
                ),
            }
            assert_eq!(
                instance.steps() - before,
                action.steps,
                "{context}: instruction count"
            );
            let stats = instance.memory_stats();
            assert_eq!(
                [
                    stats.live_bytes,
                    stats.allocated_since_sweep,
                    stats.total_allocated_bytes,
                    stats.peak_bytes
                ],
                action.memory,
                "{context}: memory"
            );
        }
        instance.close().unwrap();
        assert_eq!(instance.memory_stats().live_bytes, 0);
        assert_eq!(instance.memory_stats().allocated_since_sweep, 0);
        assert_eq!(instance.heap_stats().live_objects, 0);
    }
}
